cdtb/binfile.py

```diff
@@ -211,124 +211,110 @@ class BinType(IntEnum):
-        return f"<{self.name!r} CONTAINER({self.type.name}) {svalues}>"
+        return f"<{format_binvalue(BinType.STRUCT, self)}>"
 
     def to_serializable(self):
-        return (self.name.to_serializable(), [_to_serializable(v) for v in self.value])
+        return super().to_serializable() if self.type.h else None
 
-class BinStructField(BinField):
-    def __init__(self, hname, value):
-        super().__init__(hname)
-        self.value = value
+class BinEmbedded(BinObjectWithFieldsAndType):
+    """Embedded binary value"""
 
     def __repr__(self):
-        sfields = _repr_indent_list(self.value.fields)
-        return f"<{self.name!r} STRUCT {self.value.type!r} {sfields}>"
-
-    def to_serializable(self):
-        return (self.name.to_serializable(), self.value.to_serializable())
+        return f"<{format_binvalue(BinType.EMBEDDED, self)}>"
 
-class BinEmbeddedField(BinField):
-    def __init__(self, hname, value):
-        super().__init__(hname)
+class BinOption():
+    def __init__(self, vtype, value):
+        self.vtype = vtype
         self.value = value
 
     def __repr__(self):
-        sfields = _repr_indent_list(self.value.fields)
-        return f"<{self.name!r} EMBEDDED {self.value.type!r} {sfields}>"
+        return f"<{format_binvalue(BinType.OPTION, self)}>"
 
     def to_serializable(self):
-        return (self.name.to_serializable(), self.value.to_serializable())
+        return None if self.value is None else self.value
 
-class BinOptionField(BinField):
-    def __init__(self, hname, vtype, value):
-        super().__init__(hname)
+class BinMap():
+    def __init__(self, ktype, vtype, values: dict):
+        self.ktype = ktype
         self.vtype = vtype
-        self.value = value
+        self.values = values
 
     def __repr__(self):
-        svalue = '-' if self.value is None else f'(\n  {_repr_indent(self.value)}\n)'
-        return f"<{self.name!r} OPTION({self.vtype.name}) {svalue}>"
+        return f"<{format_binvalue(BinType.MAP, self)}>"
 
     def to_serializable(self):
-        return (self.name.to_serializable(), None if self.value is None else _to_serializable(self.value))
+        return {_to_serializable(k): _to_serializable(v) for k,v in self.values.items()}
 
-class BinMapField(BinField):
-    def __init__(self, hname, ktype, vtype, values):
-        super().__init__(hname)
-        self.ktype = ktype
-        self.vtype = vtype
-        self.value = values
+class BinField:
+    """A field is a value (possibly nested) associated to a hash."""
+
+    def __init__(self, hname, btype, value):
+        self.name = BinFieldName(hname)
+        self.type = btype
+        self.value = value
 
     def __repr__(self):
-        svalues = ''.join(f"  {k} => {_repr_indent(v)}\n" for k, v in self.value.items())
-        return f"<{self.name!r} MAP({self.ktype.name},{self.vtype.name}) {{\n{svalues}}}>"
+        return f"<{self.name!r} {format_binvalue(self.type, self.value)}>"
```

Comment:
If `self.value` is "properly typed" (i.e. it's a `BinSomething` instance) we could define `format_value(self)` (or another name) directly on `value`. This would be simpler to use: instead `format_binvalue(BinType.LIST, self)` you would have `self.format_value()`.

However, this wouldn't work for "scalar values" that don't have a dedicated `BinSomething` wrapper class.

Maybe we could defined a `format_value()` function on `BinSomething` classes and use it from the free-and-generic-function `format_binvalue()`?

I don't have a strong opinion on this. I'll let you choose what you prefer.